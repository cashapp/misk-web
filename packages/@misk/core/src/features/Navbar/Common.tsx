import { css } from "@emotion/core"
import { defaultTheme, Environment, ITheme } from "src/utilities"

export const cssNavbarHeading = (theme: ITheme) => css`
  font-size: 24px !important;
  text-decoration: none;
  text-transform: uppercase;
  letter-spacing: 0px;
  margin-right: 30px;
  color: ${theme.navbarText};
  min-width: fit-content;

  a {
    color: ${theme.navbarText} !important;
    letter-spacing: 1px;
    text-decoration: none;
    &:hover {
      color: ${theme.navbarLinkHover} !important;
      text-decoration: none;
    }
  }
`

export const cssNavbarHeadingEnvironment = (
  environment: Environment,
  theme: ITheme = defaultTheme
) =>
  css(
    cssNavbarHeading(theme),
    `
color: ${theme.environmentToColor(environment)} !important;
min-width: 0;
`
  )

export const cssMiskLink = (theme: ITheme) => css`
  color: ${theme.navbarText};
  text-decoration: none;
  &:hover {
    color: ${theme.navbarLinkHover};
    text-decoration: none;
  }
`
