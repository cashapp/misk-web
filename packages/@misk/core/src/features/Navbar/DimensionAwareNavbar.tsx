/** @jsx jsx */
import { Alignment, Navbar, NavbarGroup } from "@blueprintjs/core"
import { css, jsx } from "@emotion/core"
import * as React from "react"
import { Link } from "react-router-dom"
import { ResponsiveContainer } from "../../cssContainers"
import {
  Banner,
  HomeLink,
  IBannerExternalProps,
  IDimensionAwareProps,
  IMenuExternalProps,
  Menu,
  processNavbarItems,
  calculateNavbarItemsByScreenWidth
} from "../Navbar"
import { Environment } from "../../utilities"
import { ITheme, defaultTheme } from "src/utilities/theme"

/**
 * <DimensionAwareComponent
 *    height={this.state.height}
 *    width={this.state.width}
 *    environment={this.props.environment}
 *    environmentBannerVisible={this.props.environmentBannerVisible}
 *    environmentNavbarVisible={this.props.environmentBannerVisible}
 *    error={this.props.error}
 *    homeName={this.props.homeName}
 *    homeUrl={this.props.homeUrl}
 *    links={this.props.links}
 *    navbarItems={this.props.navbarItems}
 *    status={this.props.status}
 *    theme={this.props.theme}
 *  />
 */

export interface INavbarProps
  extends IBannerExternalProps,
    IMenuExternalProps,
    IThemeProps {
  environmentNavbarVisible?: Environment[]
  homeName?: string | Element | JSX.Element
  homeUrl?: string
  navbar_items?: Array<string | Element | JSX.Element>
  navbarItemsToDisplay?: number
}

export interface IThemeProps {
  theme?: ITheme
}

const cssNavbar = (theme: ITheme) => css`
  background-color: ${theme.navbarBackground} !important;
  padding-top: 10px !important;
  padding-bottom: 60px !important;
  position: fixed !important;
`

const cssNavbarGroup = (theme: ITheme) => css`
  font-size: 13px !important;
  font-weight: 600 !important;
  position: relative;
  padding-top: 25px;
  padding-bottom: 27px;
  @media (max-width: 870px) {
    padding-left: 60px;
  }
  @media (min-width: 992px) and (max-width: 1085px) {
    padding-left: 60px;
  }
  @media (min-width: 1200px) and (max-width: 1285px) {
    padding-left: 60px;
  }
`

export class DimensionAwareNavbar extends React.Component<
  IDimensionAwareProps & INavbarProps,
  {}
> {
  // TODO see if this can be removed
  public state = {
    isOpen: false
  }

  public render() {
    const {
      environment,
      environmentBannerVisible,
      environmentNavbarVisible,
      error,
      homeName,
      homeUrl,
      linkComponent = Link,
      links,
      menuIcon,
      menuOpenIcon,
      menuButtonAsLink,
      menuShowButton,
      navbar_items,
      navbarItemsToDisplay = null,
      status,
      theme = defaultTheme,
      width
    } = this.props
    const processedNavbarItems = processNavbarItems(
      theme,
      environment,
      environmentNavbarVisible,
      navbar_items
    )
    const numberOfNavbarItemsToDisplay =
      navbarItemsToDisplay == null
        ? calculateNavbarItemsByScreenWidth(width)
        : navbarItemsToDisplay
    return (
      <Navbar css={cssNavbar(theme)}>
        <ResponsiveContainer>
          <NavbarGroup
            align={Alignment.LEFT}
            className="bp3-dark"
            css={cssNavbarGroup(theme)}
          >
            <HomeLink
              linkComponent={linkComponent}
              homeName={homeName}
              homeUrl={homeUrl}
              theme={theme}
            />
            {processedNavbarItems.slice(0, numberOfNavbarItemsToDisplay)}
          </NavbarGroup>
        </ResponsiveContainer>
        <Menu
          processedNavbarItems={processedNavbarItems}
          error={error}
          links={links}
          linkComponent={linkComponent}
          menuIcon={menuIcon}
          menuOpenIcon={menuOpenIcon}
          menuButtonAsLink={menuButtonAsLink}
          menuShowButton={menuShowButton}
          theme={theme}
        />
        <Banner
          environment={environment}
          environmentBannerVisible={environmentBannerVisible}
          status={status}
          theme={theme}
        />
      </Navbar>
    )
  }
}
