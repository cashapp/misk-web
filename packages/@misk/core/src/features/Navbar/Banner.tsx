/** @jsx jsx */
import { css, jsx } from "@emotion/core"
import * as React from "react"
import { TextHTMLOrElementComponent } from "../../components"
import { FlexContainer, ResponsiveContainer } from "../../cssContainers"
import { Environment, ITheme, defaultTheme } from "../../utilities"
import { IThemeProps } from "./DimensionAwareNavbar"

/**
 * <Banner
 *    environment={this.props.environment}
 *    environmentBannerVisible={this.props.environmentBannerVisible}
 *    status={this.props.status}
 *  />
 */

export interface IBannerExternalProps extends IThemeProps {
  environment?: Environment
  environmentBannerVisible?: Environment[]
  status?: string | Element | JSX.Element
}

const cssNavbarBanner = (environment: Environment, theme: ITheme) => css`
  background-color: ${theme.environmentToColor(environment)} !important;
  color: ${theme.bannerText} !important;
  text-align: center;
  font-weight: 600;
  padding: 5px 10px;
  position: fixed !important;
  width: 100%;
  top: 70px;
  left: 0px;
  z-index: 1010 !important;

  a {
    font-weight: 300;
    color: ${theme.bannerText};
    text-decoration: underline;
    letter-spacing: 1px;
    &:hover {
      color: ${theme.bannerLinkHover};
      text-decoration: underline;
    }
  }
`

export class Banner extends React.Component<IBannerExternalProps, {}> {
  public render() {
    const {
      environment,
      environmentBannerVisible,
      status,
      theme = defaultTheme
    } = this.props
    if (
      environmentBannerVisible &&
      environmentBannerVisible.includes(environment)
    ) {
      return (
        <span css={cssNavbarBanner(environment, theme)}>
          <ResponsiveContainer>
            <FlexContainer>
              <TextHTMLOrElementComponent>{status}</TextHTMLOrElementComponent>
            </FlexContainer>
          </ResponsiveContainer>
        </span>
      )
    } else {
      return <div />
    }
  }
}
