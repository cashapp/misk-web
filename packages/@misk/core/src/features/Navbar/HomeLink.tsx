/** @jsx jsx */
import { css, jsx } from "@emotion/core"
import { Link } from "react-router-dom"
import { cssMiskLink, cssNavbarHeading } from "./Common"
import { IThemeProps } from "./DimensionAwareNavbar"
import { defaultTheme } from "src/utilities"

/**
 * <HomeLink
 *    homeName={this.props.homeName}
 *    homeUrl={this.props.homeUrl}
 *  />
 */

export interface IHomeLinkProps extends IThemeProps {
  homeName?: string | Element | JSX.Element
  homeUrl?: string
  linkComponent?: any
}

export const HomeLink = (props: IHomeLinkProps) => {
  const { homeName, homeUrl, theme = defaultTheme } = props
  const LinkComponent = props.linkComponent || Link
  if (homeName && homeUrl) {
    return (
      <LinkComponent
        css={css(cssMiskLink(theme), { minWidth: "fit-content" })}
        to={homeUrl}
      >
        <span css={cssNavbarHeading(theme)}>{homeName}</span>
      </LinkComponent>
    )
  } else if (homeName) {
    return <span css={cssNavbarHeading(theme)}>{homeName}</span>
  } else {
    return <span css={cssNavbarHeading(theme)}>Misk</span>
  }
}
