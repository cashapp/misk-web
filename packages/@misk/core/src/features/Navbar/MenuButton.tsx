/** @jsx jsx */
import { Button, Classes, Icon, IconName } from "@blueprintjs/core"
import { IconNames } from "@blueprintjs/icons"
import { css, jsx } from "@emotion/core"
import { Link } from "react-router-dom"
import { defaultTheme, ITheme } from "../../utilities"
import { IThemeProps } from "./DimensionAwareNavbar"

/**
 * <MenuButton
 *    handleClick={this.handleClick}
 *    isOpen={this.state.isOpen}
 *    menuIcon={this.props.menuIcon}
 *    menuOpenIcon={this.props.menuOpenIcon}
 *    menuButtonAsLink={this.props.menuButtonAsLink}
 *    menuShowButton={this.props.menuShowButton}
 *  />
 */

export interface IMenuButtonExternalProps extends IThemeProps {
  homeUrl?: string
  menuIcon?: IconName | JSX.Element | string
  menuOpenIcon?: IconName | JSX.Element | string
  menuButtonAsLink?: boolean
  menuShowButton?: boolean
}

export interface IMenuButtonProps extends IMenuButtonExternalProps {
  handleClick: () => any
  isOpen: boolean
}

const cssButton = (theme: ITheme) => css`
  background-color: ${theme.navbarBackground} !important;
  box-shadow: none !important;
  background-image: none !important;
  top: 15px;
  left: 15px;
  position: absolute;
  z-index: 1020;
`

const cssIcon = (theme: ITheme) => css`
  color: ${theme.button} !important;
  &:hover {
    color: ${theme.buttonHover};
  }
`

const cssButtonImg = css`
  max-height: 32px;
`

const ImgOrIcon = (
  props: { icon: IconName | JSX.Element | string } & IThemeProps
) => {
  const { icon, theme } = props
  if (
    typeof icon === "string" &&
    (icon.startsWith("/") || icon.startsWith("http"))
  ) {
    return <img css={cssButtonImg} src={icon as string} />
  } else {
    return <Icon css={cssIcon(theme)} iconSize={32} icon={icon as IconName} />
  }
}

const OpenCloseIcon = (
  props: {
    isOpen: boolean
    menuIcon?: IconName | JSX.Element | string
    menuOpenIcon?: IconName | JSX.Element | string
  } & IThemeProps
) => {
  const { isOpen, menuIcon, menuOpenIcon, theme } = props
  if (isOpen) {
    return <ImgOrIcon icon={menuOpenIcon} theme={theme} />
  } else {
    return <ImgOrIcon icon={menuIcon} theme={theme} />
  }
}

export const MenuButton = (props: IMenuButtonProps) => {
  const {
    handleClick,
    homeUrl,
    isOpen,
    menuIcon = IconNames.MENU,
    menuOpenIcon = IconNames.CROSS,
    menuButtonAsLink = false,
    menuShowButton = true,
    theme = defaultTheme
  } = props
  if (menuShowButton && !menuButtonAsLink) {
    return (
      <Button css={cssButton(theme)} onClick={handleClick}>
        <OpenCloseIcon
          isOpen={isOpen}
          menuIcon={menuIcon}
          menuOpenIcon={menuOpenIcon}
          theme={theme}
        />
      </Button>
    )
  } else if (menuShowButton && menuButtonAsLink) {
    return (
      <Link css={cssButton(theme)} className={Classes.BUTTON} to={homeUrl}>
        <OpenCloseIcon isOpen={false} menuIcon={menuIcon} theme={theme} />
      </Link>
    )
  } else {
    return <div />
  }
}
