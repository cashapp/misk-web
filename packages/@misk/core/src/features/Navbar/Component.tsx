import { IResizeEntry, ResizeSensor } from "@blueprintjs/core"
import reduce from "lodash/reduce"
import * as React from "react"
import {
  defaultEnvironment,
  defaultEnvironmentIndicatorsVisible
} from "../../utilities"
import { DimensionAwareNavbar, INavbarProps } from "../Navbar"
import { defaultTheme } from "src/utilities/theme"

/**
 * <Component
 *    environment={environment}
 *    environmentBannerVisible={[Environment.DEVELOPMENT]}
 *    environmentNavbarVisible={[Environment.DEVELOPMENT]}
 *    error={error}
 *    homeName={"Misk Home"}
 *    homeUrl={"/"}
 *    links={links}
 *    menuIcon={IconNames.FOLDER_CLOSE}
 *    menuOpenIcon={IconNames.FOLDER_OPEN}
 *    menuButtonAsLink={true}
 *    menuShowButton={true}
 *    navbarItems={[ "Test1", '<a href="#">Test2</>', <span key={2}>Test3</span> ]}
 *    status={"News Item"}
 *  />
 */

export interface IDimensionAwareProps {
  height: number
  width: number
}

export class Navbar extends React.Component<INavbarProps, {}> {
  public state = {
    height: 0,
    width: 0
  }

  handleResize = (entries: IResizeEntry[]) => {
    this.setState(
      reduce(
        entries,
        (dimension, e) => ({
          height: Math.max(dimension.height, e.contentRect.height),
          width: Math.max(dimension.width, e.contentRect.width)
        }),
        { height: 0, width: 0 }
      )
    )
  }

  public render() {
    const {
      environment = defaultEnvironment,
      environmentBannerVisible = defaultEnvironmentIndicatorsVisible,
      environmentNavbarVisible = defaultEnvironmentIndicatorsVisible,
      error,
      homeName,
      homeUrl,
      linkComponent,
      links,
      menuIcon,
      menuOpenIcon,
      menuButtonAsLink,
      menuShowButton,
      navbar_items,
      navbarItemsToDisplay = null,
      status,
      theme = defaultTheme
    } = this.props
    const { height, width } = this.state
    return (
      <ResizeSensor onResize={this.handleResize}>
        <DimensionAwareNavbar
          height={height}
          width={width}
          environment={environment}
          environmentBannerVisible={environmentBannerVisible}
          environmentNavbarVisible={environmentNavbarVisible}
          error={error}
          homeName={homeName}
          homeUrl={homeUrl}
          linkComponent={linkComponent}
          links={links}
          menuIcon={menuIcon}
          menuOpenIcon={menuOpenIcon}
          menuButtonAsLink={menuButtonAsLink}
          menuShowButton={menuShowButton}
          navbar_items={navbar_items}
          navbarItemsToDisplay={navbarItemsToDisplay}
          status={status}
          theme={theme}
        />
      </ResizeSensor>
    )
  }
}
