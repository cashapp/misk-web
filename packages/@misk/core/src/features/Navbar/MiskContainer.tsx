import { Spinner } from "@blueprintjs/core"
import React, { useEffect, useState } from "react"
import { ResponsiveAppContainer } from "src/cssContainers"
import { TabLoaderComponent } from "src/components"
import { INavbarProps, Navbar, choose } from "src/features/Navbar"
import {
  get,
  defaultEnvironment,
  defaultEnvironmentIndicatorsVisible,
  defaultTheme,
  IServiceMetadata,
  IDashboardMetadata
} from "src/utilities"

export interface IMiskNavbarContainerProps {
  children?: any
  dashboardMetadataUrl?: string
  loadingSpinner?: boolean
  propsOverrideRemoteData?: boolean
  serviceMetadataUrl?: string
}

export const miskDashboardMetadataUrl = (dashboardSlug: string) =>
  `/api/dashboard/${dashboardSlug}/metadata`
export const miskServiceMetadataUrl = "/api/service/metadata"
export const testAdminDashboardTabsUrl =
  "https://cashapp.github.io/misk-web/examples/data/demo/adminDashboardTabs.json"
export const testServiceMetadataUrl =
  "https://cashapp.github.io/misk-web/examples/data/demo/serviceMetadata.json"

/**
 * Container for a dashboard Navbar
 * Includes network requests to fill in NavbarProps using Misk service endpoints for
 *  * adminDashboardTabs (links), also used for TabLoaderComponent
 *  * service metadata (environment, homeName, homeUrl, navbar_items, status)
 */
export const MiskNavbarContainer = (
  props: IMiskNavbarContainerProps & INavbarProps
) => {
  const {
    dashboardMetadataUrl,
    children,
    loadingSpinner = true,
    propsOverrideRemoteData = false,
    serviceMetadataUrl,
    // INavbarProps
    environmentBannerVisible = defaultEnvironmentIndicatorsVisible,
    environmentNavbarVisible = defaultEnvironmentIndicatorsVisible,
    error,
    homeName,
    homeUrl,
    linkComponent,
    links,
    menuIcon,
    menuOpenIcon,
    menuButtonAsLink,
    menuShowButton,
    navbar_items,
    navbarItemsToDisplay = null,
    status,
    theme = defaultTheme
  } = props

  // State initialization
  const [loading, setLoading] = useState(false)
  const [dashboardMetadata, setDashboardMetadata] = useState(
    {} as IDashboardMetadata
  )
  const [serviceMetadata, setServiceMetadata] = useState({} as IServiceMetadata)

  // Network calls to get dashboardMetadata and serviceMetadata
  const getDashboardMetadata = async () => {
    const { data, error } = await get(dashboardMetadataUrl)
    if (error) {
      console.error(`Failed to GET ${dashboardMetadataUrl}`, error)
      setLoading(true)
    } else {
      setDashboardMetadata(data.dashboardMetadata)
      setLoading(false)
    }
  }

  const getServiceMetadata = async () => {
    const { data, error } = await get(serviceMetadataUrl)
    if (error) {
      console.error(`Failed to GET ${serviceMetadataUrl}`, error)
      setLoading(true)
    } else {
      setServiceMetadata(data.serviceMetadata)
    }
  }

  useEffect(() => {
    if (dashboardMetadataUrl) {
      setLoading(true)
      getDashboardMetadata()
    }
    if (serviceMetadataUrl) {
      setLoading(true)
      getServiceMetadata()
    }
  }, [])

  return (
    <>
      <Navbar
        environment={serviceMetadata.environment || defaultEnvironment}
        environmentBannerVisible={environmentBannerVisible}
        environmentNavbarVisible={environmentNavbarVisible}
        error={error}
        homeName={choose(
          propsOverrideRemoteData,
          homeName,
          serviceMetadata.app_name
        )}
        homeUrl={choose(
          propsOverrideRemoteData,
          homeUrl,
          dashboardMetadata.home_url
        )}
        linkComponent={linkComponent}
        links={choose(propsOverrideRemoteData, links, dashboardMetadata.tabs)}
        menuIcon={menuIcon}
        menuOpenIcon={menuOpenIcon}
        menuButtonAsLink={menuButtonAsLink}
        menuShowButton={menuShowButton}
        navbar_items={choose(
          propsOverrideRemoteData,
          navbar_items,
          dashboardMetadata.navbar_items
        )}
        navbarItemsToDisplay={navbarItemsToDisplay}
        status={choose(
          propsOverrideRemoteData,
          status,
          dashboardMetadata.navbar_status
        )}
        theme={theme}
      />
      <ResponsiveAppContainer>
        {loadingSpinner && loading && <Spinner />}
        <TabLoaderComponent tabs={dashboardMetadata.tabs} />
        {children}
      </ResponsiveAppContainer>
    </>
  )
}
