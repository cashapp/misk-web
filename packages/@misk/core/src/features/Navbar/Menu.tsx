/** @jsx jsx */
import { Collapse } from "@blueprintjs/core"
import { css, jsx } from "@emotion/core"
import { chain, sortBy } from "lodash"
import * as React from "react"
import { Link } from "react-router-dom"
import { ErrorCalloutComponent } from "../../components"
import { FlexContainer, ResponsiveContainer } from "../../cssContainers"
import { IDashboardTab, defaultTheme, ITheme } from "../../utilities"
import { cssMiskLink } from "./Common"
import { MenuButton, IMenuButtonExternalProps } from "./MenuButton"
import { IThemeProps } from "./DimensionAwareNavbar"

/**
 * <Menu
 *    error={this.props.error}
 *    links={this.props.links}
 *    processedNavbarItems={this.props.processedNavbarItems}
 *    ...IMenuButtonExternalProps...
 *  />
 */

export interface IMenuExternalProps
  extends IMenuButtonExternalProps,
    IThemeProps {
  error?: any
  links?: IDashboardTab[]
  linkComponent?: any
}

export interface IMenuProps extends IMenuExternalProps {
  processedNavbarItems?: JSX.Element[]
}

const cssCollapse = (theme: ITheme) => css`
  background-color: ${theme.navbarBackground};
  display: block;
  margin: 60px -20px 0 -20px;
`

const cssMenu = css`
  min-height: 250px;
  padding: 50px 0px;
  @media (max-width: 768px) {
    padding: 50px 20px;
  }
  overflow-y: scroll;
  max-height: 100vh;
`

const cssMenuNavbarItems = css`
  display: inline-block;
`

const cssMenuLinks = css`
  padding-bottom: 35px;
`

const cssMenuLink = (theme: ITheme) => css`
  font-size: 16px;
  flex-basis: 300px;
  padding: 5px 0;
  color: ${theme.navbarText};
`

const cssMenuCategory = (theme: ITheme) => css`
  font-size: 24px;
  color: ${theme.categoryText};
  letter-spacing: 0px;
  display: block;
`

const cssMenuDivider = (theme: ITheme) => css`
  border-color: ${theme.categoryText};
  margin: 5px 0 10px 0;
`

const MenuHeading = (props: { categoryName: string } & IThemeProps) => {
  const { categoryName, theme = defaultTheme } = props
  if (categoryName === "undefined") {
    return <span />
  } else {
    return (
      <div>
        <span css={css(cssMenuCategory(theme))}>{categoryName}</span>
        <hr css={cssMenuDivider(theme)} />
      </div>
    )
  }
}

const MenuCategory = (
  props: {
    categoryName: string
    categoryLinks: IDashboardTab[]
    handleClick: () => void
    LinkComponent: any
  } & IThemeProps
) => {
  const {
    categoryName,
    categoryLinks,
    handleClick,
    LinkComponent,
    theme = defaultTheme
  } = props
  return (
    <div>
      <MenuHeading categoryName={categoryName} />
      <FlexContainer css={cssMenuLinks}>
        {categoryLinks &&
          categoryLinks.map((link: IDashboardTab) => {
            const cssProps = css(cssMiskLink(theme), cssMenuLink(theme))
            if (link.url_path_prefix.startsWith("http")) {
              return (
                <a
                  css={cssProps}
                  key={link.slug}
                  onClick={handleClick}
                  href={link.url_path_prefix}
                >
                  {link.name}
                </a>
              )
            } else {
              return (
                <LinkComponent
                  css={cssProps}
                  key={link.slug}
                  onClick={handleClick}
                  to={link.url_path_prefix}
                >
                  {link.name}
                </LinkComponent>
              )
            }
          })}
      </FlexContainer>
    </div>
  )
}

export class Menu extends React.Component<IMenuProps, {}> {
  public state = {
    isOpen: false
  }

  public render() {
    const { isOpen } = this.state
    const {
      error,
      links,
      linkComponent = Link,
      processedNavbarItems,
      menuIcon,
      menuOpenIcon,
      menuButtonAsLink,
      menuShowButton,
      theme = defaultTheme
    } = this.props
    return (
      <div>
        <MenuButton
          handleClick={this.handleClick}
          isOpen={isOpen}
          menuIcon={menuIcon}
          menuOpenIcon={menuOpenIcon}
          menuButtonAsLink={menuButtonAsLink}
          menuShowButton={menuShowButton}
          theme={theme}
        />
        <div css={cssCollapse(theme)}>
          <Collapse isOpen={isOpen} keepChildrenMounted={true}>
            <div css={cssMenu}>
              <ResponsiveContainer>
                <div css={cssMenuNavbarItems}>
                  <FlexContainer>
                    {processedNavbarItems &&
                      processedNavbarItems.map((item, index) => (
                        <span key={index} onClick={this.handleClick}>
                          {item}
                        </span>
                      ))}
                  </FlexContainer>
                </div>
                {links ? (
                  this.renderMenuCategories(links, linkComponent, theme)
                ) : (
                  <ErrorCalloutComponent error={error} />
                )}
              </ResponsiveContainer>
            </div>
          </Collapse>
        </div>
      </div>
    )
  }

  private renderMenuCategories(
    links: IDashboardTab[],
    LinkComponent: any,
    theme: ITheme
  ) {
    const categories: Array<[string, IDashboardTab[]]> = Object.entries(
      // sort and group array of links by category string
      // and sort each category's links by lower case tab name string
      chain(links)
        .sortBy("category")
        .groupBy("category")
        .mapValues(category =>
          sortBy(category, (tab: IDashboardTab) => tab.name.toLowerCase())
        )
        .value()
    )
    return (
      categories &&
      categories.map(([categoryName, categoryLinks], index) => (
        <MenuCategory
          categoryName={categoryName}
          categoryLinks={categoryLinks}
          key={`${categoryName}-${index}`}
          handleClick={this.handleClick}
          LinkComponent={LinkComponent}
          theme={theme}
        />
      ))
    )
  }

  private handleClick = () => {
    this.setState({ ...this.state, isOpen: !this.state.isOpen })
  }
}
