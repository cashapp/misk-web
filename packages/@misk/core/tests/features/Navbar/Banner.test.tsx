const { getByText } = require("@testing-library/dom")
import { cleanup, render } from "@testing-library/react"
import { matchers } from "jest-emotion"
import React from "react"
import { Banner } from "src/features/Navbar"
import { Environment, defaultTheme, color } from "src"

expect.extend(matchers)

describe("Navbar", () => {
  afterEach(cleanup)
  it("Banner can render empty", () => {
    const { asFragment } = render(<Banner />)
    expect(asFragment()).toMatchSnapshot()
  })
  it("Banner can render with props but banner is not set as visible", () => {
    const { asFragment } = render(
      <Banner
        environment={Environment.DEVELOPMENT}
        environmentBannerVisible={[]}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
  it("Banner can render with props and banner is set to visible", () => {
    const { asFragment } = render(
      <Banner
        environment={Environment.DEVELOPMENT}
        environmentBannerVisible={[Environment.DEVELOPMENT]}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
  it("Banner can render with props, banner is set to visible, and status is displayed", () => {
    const status = "Alert: Visible Status!"
    const { asFragment, container } = render(
      <Banner
        environment={Environment.DEVELOPMENT}
        environmentBannerVisible={[Environment.DEVELOPMENT]}
        status={status}
      />
    )
    // TODO add test that queries the status code and checks that the exact status text is present in snapshot
    expect(getByText(container, status)).toBeDefined()
    expect(asFragment()).toMatchSnapshot()
  })
  it("Banner renders with theme colors", () => {
    const status = "Alert: Visible Status!"
    const { asFragment } = render(
      <Banner
        environment={Environment.DEVELOPMENT}
        environmentBannerVisible={[Environment.DEVELOPMENT]}
        status={status}
        theme={{ ...defaultTheme, bannerText: color.purple }}
      />
    )
    // TODO add test that queries the status code and checks that the exact status text is present in snapshot
    // TODO add tests that theme values actually propogate
    expect(asFragment()).toMatchSnapshot()
  })
})
