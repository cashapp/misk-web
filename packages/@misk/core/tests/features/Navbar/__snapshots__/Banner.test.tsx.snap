// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar Banner can render empty 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`Navbar Banner can render with props and banner is set to visible 1`] = `
<DocumentFragment>
  .emotion-2 {
  background-color: #49a4d5 !important;
  color: #eceef1 !important;
  text-align: center;
  font-weight: 600;
  padding: 5px 10px;
  position: fixed !important;
  width: 100%;
  top: 70px;
  left: 0px;
  z-index: 1010 !important;
}

.emotion-2 a {
  font-weight: 300;
  color: #eceef1;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
}

.emotion-2 a:hover {
  color: white;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-1 {
  margin: 0 auto;
}

@media (min-width:768px) {
  .emotion-1 {
    width: 750px;
  }
}

@media (min-width:992px) {
  .emotion-1 {
    width: 970px;
  }
}

@media (min-width:1200px) {
  .emotion-1 {
    width: 1170px;
  }
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.emotion-0 * {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<span
    class="emotion-2"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-0"
      >
        <span />
      </div>
    </div>
  </span>
</DocumentFragment>
`;

exports[`Navbar Banner can render with props but banner is not set as visible 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`Navbar Banner can render with props, banner is set to visible, and status is displayed 1`] = `
<DocumentFragment>
  .emotion-2 {
  background-color: #49a4d5 !important;
  color: #eceef1 !important;
  text-align: center;
  font-weight: 600;
  padding: 5px 10px;
  position: fixed !important;
  width: 100%;
  top: 70px;
  left: 0px;
  z-index: 1010 !important;
}

.emotion-2 a {
  font-weight: 300;
  color: #eceef1;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
}

.emotion-2 a:hover {
  color: white;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-1 {
  margin: 0 auto;
}

@media (min-width:768px) {
  .emotion-1 {
    width: 750px;
  }
}

@media (min-width:992px) {
  .emotion-1 {
    width: 970px;
  }
}

@media (min-width:1200px) {
  .emotion-1 {
    width: 1170px;
  }
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.emotion-0 * {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<span
    class="emotion-2"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-0"
      >
        <span>
          Alert: Visible Status!
        </span>
      </div>
    </div>
  </span>
</DocumentFragment>
`;

exports[`Navbar Banner renders with theme colors 1`] = `
<DocumentFragment>
  .emotion-1 {
  margin: 0 auto;
}

@media (min-width:768px) {
  .emotion-1 {
    width: 750px;
  }
}

@media (min-width:992px) {
  .emotion-1 {
    width: 970px;
  }
}

@media (min-width:1200px) {
  .emotion-1 {
    width: 1170px;
  }
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.emotion-0 * {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.emotion-2 {
  background-color: #49a4d5 !important;
  color: #7250a1 !important;
  text-align: center;
  font-weight: 600;
  padding: 5px 10px;
  position: fixed !important;
  width: 100%;
  top: 70px;
  left: 0px;
  z-index: 1010 !important;
}

.emotion-2 a {
  font-weight: 300;
  color: #7250a1;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
}

.emotion-2 a:hover {
  color: white;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<span
    class="emotion-2"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-0"
      >
        <span>
          Alert: Visible Status!
        </span>
      </div>
    </div>
  </span>
</DocumentFragment>
`;
